File: app/src/main/java/com/bove/martin/pexel/util/GsonDeserializador.java
package com.bove.martin.pexel.util;

import com.bove.martin.pexel.model.Foto;
import com.google.gson.JsonArray;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonDeserializer;
import com.google.gson.JsonElement;
import com.google.gson.JsonParseException;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by Martín Bove on 28/05/2018.
 * E-mail: mbove77@gmail.com
 */
public class GsonDeserializador implements JsonDeserializer<List<Foto>> {
    List<Foto> fotos = new ArrayList<Foto>();
    @Override
    public List<Foto> deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) throws JsonParseException {
        JsonArray photos = json.getAsJsonObject().get("photos").getAsJsonArray();

        for (int i=0; i < photos.size(); i++) {
            int id = photos.get(i).getAsJsonObject().get("id").getAsInt();
            int width =  photos.get(i).getAsJsonObject().get("width").getAsInt();
            int height = photos.get(i).getAsJsonObject().get("height").getAsInt();
            String url = photos.get(i).getAsJsonObject().get("url").getAsString();
            String photographer = photos.get(i).getAsJsonObject().get("photographer").getAsString();
            String original = photos.get(i).getAsJsonObject().get("src").getAsJsonObject().get("original").getAsString();
            String large = photos.get(i).getAsJsonObject().get("src").getAsJsonObject().get("large").getAsString();
            String large2x = photos.get(i).getAsJsonObject().get("src").getAsJsonObject().get("large2x").getAsString();
            String medium = photos.get(i).getAsJsonObject().get("src").getAsJsonObject().get("medium").getAsString();
            String small = photos.get(i).getAsJsonObject().get("src").getAsJsonObject().get("small").getAsString();
            String portrait = photos.get(i).getAsJsonObject().get("src").getAsJsonObject().get("portrait").getAsString();
            String landscape = photos.get(i).getAsJsonObject().get("src").getAsJsonObject().get("landscape").getAsString();
            String tiny = photos.get(i).getAsJsonObject().get("src").getAsJsonObject().get("tiny").getAsString();

            Foto currentFoto = new Foto(id, width, height, url, photographer, original, large, large2x, medium, small, portrait, landscape, tiny);

            if(!fotos.contains(currentFoto))
                fotos.add(currentFoto);
        }

        return fotos;
    }
}

File: app/src/main/java/com/bove/martin/pexel/data/retrofit/GsonDeserializador.kt
package com.bove.martin.pexel.data.retrofit

import com.bove.martin.pexel.data.model.Foto
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonParseException
import java.lang.reflect.Type
import java.util.*

/**
 * Created by Martín Bove on 28/05/2018.
 * E-mail: mbove77@gmail.com
 */
class GsonDeserializador : JsonDeserializer<List<Foto>> {
    var fotos: MutableList<Foto> = ArrayList()
    @Throws(JsonParseException::class)
    override fun deserialize(json: JsonElement, typeOfT: Type, context: JsonDeserializationContext): List<Foto> {
        val photos = json.asJsonObject["photos"].asJsonArray
        for (i in 0 until photos.size()) {
            val id = photos[i].asJsonObject["id"].asInt
            val width = photos[i].asJsonObject["width"].asInt
            val height = photos[i].asJsonObject["height"].asInt
            val url = photos[i].asJsonObject["url"].asString
            val photographer = photos[i].asJsonObject["photographer"].asString
            val original = photos[i].asJsonObject["src"].asJsonObject["original"].asString
            val large = photos[i].asJsonObject["src"].asJsonObject["large"].asString
            val large2x = photos[i].asJsonObject["src"].asJsonObject["large2x"].asString
            val medium = photos[i].asJsonObject["src"].asJsonObject["medium"].asString
            val small = photos[i].asJsonObject["src"].asJsonObject["small"].asString
            val portrait = photos[i].asJsonObject["src"].asJsonObject["portrait"].asString
            val landscape = photos[i].asJsonObject["src"].asJsonObject["landscape"].asString
            val tiny = photos[i].asJsonObject["src"].asJsonObject["tiny"].asString
            val currentFoto = Foto(id, width, height, url, photographer, original, large, large2x, medium, small, portrait, landscape, tiny)
            if (!fotos.contains(currentFoto)) fotos.add(currentFoto)
        }
        return fotos
    }
}